airbyte-integrations/connectors/source-mailchimp/source_mailchimp/client.py

```diff
@@ -48,25 +53,69 @@ def health_check(self):
         return streams
 
-    def lists(self):
-        limit = self._client.lists.all(count=1)["total_items"]
+    def lists(self, state: DefaultDict[str, any] = None) -> Generator[AirbyteMessage, None, None]:
+        cursor_field = 'date_created'
+        stream_name = self._LISTS
+        date_created = self._get_cursor_or_none(state, stream_name, cursor_field)
 
+        default_params = {'since_date_created': date_created, 'sort_field': cursor_field, 'sort_dir': 'ASC'}
         offset = 0
-        while offset < limit:
-            for mc_list in self._client.lists.all(count=self.PAGINATION, offset=offset)["lists"]:
-                yield mc_list
+        max_date_created = date_created
+        done = False
+        while not done:
+            lists_response = self._client.lists.all(dict(default_params, count=self.PAGINATION, offset=offset))["lists"]
+            for mc_list in lists_response:
+                list_created_at = parser.isoparse(mc_list[cursor_field])
+                max_date_created = max(max_date_created, list_created_at) if max_date_created else list_created_at
+                yield self._record(stream=self._LISTS, data=mc_list)
+
+            if max_date_created:
+                state[self._LISTS][cursor_field] = max_date_created
+                yield self._state(state)
+
+            done = len(lists_response) < self.PAGINATION
             offset += self.PAGINATION
 
-    def campaigns(self):
-        limit = self._client.campaigns.all(count=1)["total_items"]
+    def campaigns(self, state: DefaultDict[str, any]) -> Generator[AirbyteMessage, None, None]:
+        cursor_field = 'create_time'
+        stream_name = self._CAMPAIGNS
+        create_time = self._get_cursor_or_none(state, stream_name, cursor_field)
 
+        default_params = {'since_create_time': create_time, 'sort_field': cursor_field, 'sort_dir': 'ASC'}
         offset = 0
-        while offset < limit:
-            for mc_campaigns in self._client.campaigns.all(count=self.PAGINATION, offset=offset)["campaigns"]:
-                yield mc_campaigns
+        max_create_time = create_time
+        done = False
+        while not done:
+            campaigns_response = self._client.campaigns.all(dict(default_params, count=self.PAGINATION, offset=offset))
+            for campaign in campaigns_response["campaigns"]:
+                campaign_created_at = parser.isoparse(campaign[cursor_field])
+                max_create_time = max(max_create_time, campaign_created_at) if max_create_time else campaign_created_at
+                yield self._record(stream=stream_name, data=campaign)
+
+            if max_create_time:
+                state[stream_name][cursor_field] = max_create_time
+                yield self._state(state)
+
+            done = len(campaigns_response) < self.PAGINATION
             offset += self.PAGINATION
+
+    @staticmethod
+    def _get_cursor_or_none(state: DefaultDict[str, any], stream_name: str, cursor_name: str) -> any:
```

Comment:
How much of this code can become a reusable helper?

Reply:
I think we can DRY emitting record/state messages, and potentially all state retrieval/update methods into cross-integration helpers. 

Also, I think we can DRY up the REST API access for this integration (after adding child endpoints). Still unclear if we can have a reusable REST API helper _across_ integrations, but that would be amazing.